.adr/0016-closing-go-routines.md

```diff
@@ -0,0 +1,128 @@
+
+**Note:** I use the term struct as a shorthand for a struct with a long-running go-routine, like the examples above.
+
+When a struct's `Close` function is called, it should block and not return until:
+
+- all go-routines it owns stop executing.
+- any closeable resources it owns are closed.
+
+By enforcing these constraints a running go-routine can be guaranteed that it's parent struct is in a "running" state. This rules out a large class of race conditions and errors such as a go-routine attempting to use a parent struct's resource that has been closed.
+
+To enforce these constraints we should follow this pattern in a struct's `Close` function:
+
+1. Signal any go-routines we own to exit.
+2. Wait until all go-routines have completed execution.
+3. Close any resources it owns.
+
+### Step 1: Signal go-routines to exit
+
+Long-running `go-routines` need some kind of trigger to stop executing. A common and simple pattern we often use is simply closing the chan the go-routine is consuming from.
+
+```golang
+	toRoutine := make(chan int)
+
+	go func() {
+		for v := range toRoutine {
+			doSomething(v)
+		}
+		// This logic runs once the channel is closed
+		doCleanup()
+	}()
+
+	// This closes the channel and signals the go routine to exit
+	close(toRoutine)
+
+```
+
+This works, however a [cancelable context](https://cs.opensource.google/go/go/+/go1.20.5:src/context/context.go;l=238) provides some benefits over this:
+
+- A buffered chan will only get closed once it's buffer is read. This means a go-routine will read all the buffered entries before it finishes executing. By using a context we can halt the execution almost immediately.
+- It makes go-routine cleanup logic explicit and easy to see. It's now just a case statement for `ctx.Done`.
+- Minor but it would allow us to use other context features, such as timeouts, in the future.
+
+Due to these benefits, and the limited use of go-routines, we should update our structs to use a cancelable context to signal to a go-routine to stop executing.
+
+```golang
+	toRoutine := make(chan int)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case v := <-toRoutine:
+				doSomething(v)
+			case <-ctx.Done():
+				doCleanup()
+			}
+		}
+	}()
+
+	// This triggers the goroutine to exit
+	cancel()
+```
+
+### 2: Wait until all go-routines have completed execution.
+
+After we have signalled our go-routines to exit we should wait for them to complete. The easiest way to accomplish this is with a `sync.WaitGroup`
+
+```golang
+wg := sync.WaitGroup{}
+	toRoutine := make(chan int)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg.Add(1)
+	go func() {
+		for {
+			select {
+			case v := <-toRoutine:
+				doSomething(v)
+			case <-ctx.Done():
+				doCleanup()
+				wg.Done()
+			}
+		}
+	}()
+
+	// This triggers the goroutine to exit
+	cancel()
+
+	// This blocks until the goroutine has exited
+	wg.Wait()
+```
+
+### Step 3: Close Resources
```

Comment:
>Do we have a good justification for doing this after the goroutine cleanups steps? Why not before? I feel like there is something slightly missing around thinking about the dependency graph and the order in which things are closed.

Yup, this protects the `go-routine` from accessing a struct resource that has been closed. If the go-routine can still be running after we close resources, it has to deal with the scenario that things it wants to use are closed. By enforcing that the go-routine gets closed first we avoid the situtation of the `go-routine`

A good example of this is the API client sending out notifications in the [handleEngineEvents go-routine](https://github.com/statechannels/go-nitro/blob/71ea5455a35824c06abffc0b51ca02f346e1f66f/client/client.go#L94). If we close all resource first we'd close `c.channelNotifier` before attempting to stop the go-routine. However `handleEngineEvents` could still be running and attempt to use the closed `c.channelNotifier` (Closing `c.engine` would close the chan `handleEngineEvents` reads from, but it would process the buffer first)

It's true that in some cases it could be safe to shutdown a go-routine before closing a structs resources since the go-routine doesn't use those resources. However that requires reasoning through the `go-routine` dependencies and could lead to brittle code. For example say we have some go-routine `doFoo` in the API client that doesn't use the `engine`, so we close it before closing the engine. However now there's a footgun in `doFoo`, if someone modifies it use the engine it could introduce a race condition where `doFoo` tries to use the engine after its been closed. 

Due to the nature of go-routines I think it will always be safe to stop the go-routines first. A go-routine can depend on struct resources, by referencing and use the struct resources in the go-routine. I don't think there's any way the inverse would be true, where a struct has a reference to the go-routine execution and could crash if it stops.


>I also feel like cleanup the go routines is just a special case of closing resources. Shouldn't we extend the principle of "wait for the goroutine to exit" to "wait for the child struct / resource to be closed"?

I think this principle applies in to structs as well, it's just a lot easier to do. In general `"wait for the child struct / resource to be closed"?` means calling `Close` on the child struct. The child struct is the one responsible for doing cleanup and then returning from `Close` (at least in our codebase).  I think you're right that go-routines are a special case of closing resources; however I view this ADR as addressing that special case and some of it's quirks.